src/Xrefcheck/Verify.hs

```diff
@@ -445,94 +469,113 @@ verifyReference
+              pure res
+            -- Fixable, retry
+            Just (mbCurrentRetryAfter, counterModifier) -> do
+              let toSeconds = \case
+                    Seconds s -> s
+                    -- Calculates the seconds left until @Retry-After@ date.
+                    -- Defaults to 0 if the date has already passed.
+                    Date date | utcTimeToTimeSecond date >= now -> utcTimeToTimeSecond date -:- now
+                    _ -> sec 0
+
+              let currentRetryAfter = fromMaybe (ncDefaultRetryAfter cNetworking) $
+                    fmap toSeconds mbCurrentRetryAfter
+
+              let moveProgressOnFixable = whenFunc (not errorsOccured_) $
+                    composeFuncList
+                    [ incProgressFixableErrors
+                    , incProgress
+                    ]
+              modifyProgressRef now (Just currentRetryAfter) moveProgressOnFixable
+              threadDelay currentRetryAfter
+              retryVerification
+                (counterModifier rc)
+                resIO
+
+    modifyProgressRef :: Time Second -> Maybe (Time Second) -> (Progress Int -> Progress Int) -> IO ()
+    modifyProgressRef now mbRetryAfter moveProgress = atomicModifyIORef' progressRef $ \VerifyProgress{..} ->
+      ( if isExternal $ locationType rLink
+        then VerifyProgress{ vrExternal =
+          let vrExternalAdvanced = moveProgress vrExternal
+          in case mbRetryAfter of
+             Just retryAfter -> case pTaskTimestamp vrExternal of
+                    Just (TaskTimestamp ttc start)
+                      | retryAfter +:+ now <= ttc +:+ start -> vrExternalAdvanced
+                    _ -> setTaskTimestamp retryAfter now vrExternalAdvanced
+             Nothing -> vrExternalAdvanced, .. }
+        else VerifyProgress{ vrLocal = moveProgress vrLocal, .. }
+      , ()
+      )
+
+    addDomainIf429 :: IORef (S.Set DomainName) -> VerifyError -> IO (S.Set DomainName)
+    addDomainIf429 setRef err = atomicModifyIORef' setRef $ \s ->
+      (\x -> (x, x)) $ case err of
+      ExternalHttpTooManyRequests _ mbDomain ->
+        maybe s (flip S.insert s) mbDomain
+      _ -> s
+
+    decideWhetherToRetry ::
+      S.Set DomainName ->
+      RetryCounter ->
+      VerifyError ->
+      Maybe (Maybe RetryAfter, RetryCounter -> RetryCounter)
+    decideWhetherToRetry setOfReturned429 rc = \case
+      ExternalHttpTooManyRequests retryAfter _
+        | totalRetriesNotExceeded -> Just (retryAfter, incTotalCounter)
+      ExternalHttpTimeout (Just domain)
+        | totalRetriesNotExceeded && timeoutRetriesNotExceeded ->
+          -- If a given domain ever returned 429 error, we assume that getting timeout from
+          -- the domain can be considered as a 429-like error, and hence we retry.
+          -- If there was no 429 responses from this domain, then getting timeout from
+          -- it probably means that this site is not working at all.
```

Comment:
Very good :+1: 

And I would also mention that always remains a possibility that we just didn't get the response in time, but this issue cannot be mitigated at this level.

Reply:
Great, thanks :+1: